### PR TITLE
ValidatingAdmissionPolicy: typed variables support.

### LINKED ISSUE #120239: ValidatingAdmissionPolicy: variable expression type not recognized as return type in expression that evaluates the variable
### What would you like to be added?

Variables field is added to ValidatingAdmissionPolicy in 1.28. Currently the usage of variables would expect a operator associated. 
e.g. (thanks to Mariam Fahmy for raising the issue)
```
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingAdmissionPolicy
metadata:
  name: "chech-deployment-labels"
spec:
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  variables:
    - name: environment
      expression: "has(object.metadata.labels) && 'env' in object.metadata.labels && object.metadata.labels['env'] == 'prod'"
  validations:
    - expression: "variables.environment"
      message: "Deployment labels must be env=prod"
```
Above will fail with error: 
```
The ValidatingAdmissionPolicy "chech-deployment-labels" is invalid: spec.validations[0].expression: Invalid value: "variables.environment": must evaluate to bool
```
Meanwhile the `- expression: "variables.environment == true"` will pass.

The reason is when variables be evaluated to non dyn type(like bool or string), it will require special handling(like casting) since we use dyn type everywhere. 

### Why is this needed?

It might confuse user.

### PR DESCRIPTION
**Pull request by jiahuif**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds tracking of the type of each compiled composition variable for further compilation in following variables and validation expressions.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/120239

#### Does this PR introduce a user-facing change?

```release-note
ValidatingAdmissionPolicy now preserves types of composition variables, and raise type-related errors early.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by cici37** [comment]
/lgtm
This approach looks good to me. Please also cherry-pick the changes into 1.28 when it's merged. Thank you!
/approve

/assign @jpbetz 
for approval. Thanks!

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go [hunk 1]
```diff
@@ -196,3 +197,39 @@ func (a *variableAccessor) Callback(_ *lazy.MapValue) ref.Val {
 	}
 	return v
 }
+
+func convertCelTypeToDeclType(celType *cel.Type) *apiservercel.DeclType {
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:201]: Is it possible to capture type params?

Comment (jiahuif) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/composition.go:201]: Yes it is! Added support for type params.

However, note that the object is still Dyn so any access to its field remains Dyn.

**Comment by cici37** [comment]
@jiahuif Would you open the cherry pick for this one as well? Thank you!

**Pull request merged**